src/irmin-pack/layered/ext_layered.ml

```diff
@@ -231,14 +231,13 @@ struct
         end
 
         let batch t f =
-          Lwt_mutex.with_lock t.batch_lock @@ fun () ->
-          Contents.CA.batch t.contents (fun contents ->
-              Node.CA.batch t.node (fun node ->
-                  Commit.CA.batch t.commit (fun commit ->
-                      let contents : 'a Contents.t = contents in
-                      let node : 'a Node.t = (contents, node) in
-                      let commit : 'a Commit.t = (node, commit) in
-                      f contents node commit)))
+          Contents.CA.batch t.contents @@ fun contents ->
```

Comment:
Moving this lock, means that "incomplete" commits can be added during one cycle of newies. I'll add a test to expose the issue.

Reply:
Something that I don't understand is that it's already currently possible to use the "private" functions to add entries to the pack files, even if the lock is taken. Relying on `batch` for this kind of operation is not precise enough - use-code will use `batch` to schedule write barriers (as they always end-up with a flush) it's not at the right level for what you are trying to achieve (I think). 

Will try to think of a better solution.

Reply:
I added a [test on top of this PR](https://github.com/icristescu/irmin/commit/ce5830a0706eff270527dc164e1b4fb266677745) in which one new commit is not added entirely during the freeze - the contents are added during the freeze as newies, while the nodes and the commit are added after the freeze finished. 

We need to ensure that commits are atomic during the freeze - the main thread and the freeze thread cannot interleave during one commit, otherwise some objects are lost. This is because newies are copied to the next upper in [batches of commits](https://github.com/mirage/irmin/blob/master/src/irmin-pack/layered/ext_layered.ml#L586).

But we see in the benchmarks that @zshipko run that the batch lock is blocking the main thread for too long
`Longests blocks: [1 long block in "wait for batch lock" of 4.028s; 8 long blocks in "copy to next upper" of ~18.153s (max:1min48s); 1 long block in "copy newies (last)" of 1.447s].`
 so we need to remove the batch lock. Maybe revert some of the changes of https://github.com/mirage/irmin/pull/1205? This makes the freeze longer but we are not blocking the main thread. 

@zshipko  can you have a look at the test? I had to use a few locks to ensure the right interleaving between the freeze thread and the main thread and I’m worried it is not super clear.